File: 1/2020.09.21/84.Largest Rectangle in Histogram/java/solution.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        Stack<Integer> stack = new Stack<>();
        int maxArea = 0;
        for (int i = 0; i < heights.length + 1; i++) {
            int height = i < heights.length ? heights[i] : 0;
            while (!stack.isEmpty() && height < heights[stack.peek()]) {
                int h = heights[stack.pop()];
                int w = stack.isEmpty() ? i : (i-stack.peek()-1);
                maxArea = Math.max(maxArea, w*h);
            }
            stack.push(i);
        }
        return maxArea;
    }
}

File: 1/2020.09.21/84.Largest Rectangle in Histogram/python/solution.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        heights.append(0)
        stack = [-1]
        max_area = 0
        for i, height in enumerate(heights):
            while height < heights[stack[-1]]:
                h = heights[stack.pop()]
                max_area = max(max_area, (i-stack[-1]-1) * h)
            stack.append(i)
        heights.pop()
        return max_area
